File: advent_of_code/day5/part1/main.py
from typing import List

from advent_of_code.helpers.helpers import read_file_line_by_line

class Seat(object):
    def __init__(self, row: int = None, column: int = None) -> None:
        super().__init__()
        self.row: int = row
        self.column: int = column
        self.seat_id = self.get_seat_id()

    def get_seat_id(self) -> int:
        return (self.row * 8) + self.column


def convert(line: str) -> int:
    s: str = line.replace("F", "0").replace("B", "1").replace("L", "0").replace("R", "1")
    n: int = int(s, base=2)
    return n


def create_seat(line: str) -> Seat:
    row_string: str = line[:7]
    column_string: str = line[-3:]
    s: Seat = Seat(row=convert(row_string), column=convert(column_string))
    return s


def main(filepath: str) -> int:
    lines: List[str] = read_file_line_by_line(filepath=filepath)
    seats: List[Seat] = [create_seat(line=line) for line in lines]
    sorted_seat: Seat = sorted(seats, key=lambda s: s.seat_id, reverse=True)[0]
    return sorted_seat.seat_id

File: advent_of_code/day5/part1/main.go
package part1

import (
	"advent-of-code/advent_of_code/helpers"
	"fmt"
	"sort"
	"strconv"
	"strings"
)

type Seat struct {
	Row    int
	Column int
	SeatID int
}

func Convert(line string) int {
	line = strings.ReplaceAll(line, "F", "0")
	line = strings.ReplaceAll(line, "B", "1")
	line = strings.ReplaceAll(line, "L", "0")
	line = strings.ReplaceAll(line, "R", "1")
	n, err := strconv.ParseInt(line, 2, 64)

	if err != nil {
		return 0
	}

	return int(n)

}

func CreateSeat(line string) Seat {

	row := Convert(line[:7])
	column := Convert(line[7:])
	seatID := row*8 + column

	s := Seat{row, column, seatID}

	return s

}

func Main(filepath string) int {

	lines, err := helpers.ReadFileLineByLine(filepath)

	if err != nil {
		fmt.Println(err)
	}
	seats := make([]Seat, 0)

	for _, i := range lines {
		seats = append(seats, CreateSeat(i))
	}

	sort.Slice(seats, func(i, j int) bool {
		return seats[i].SeatID > seats[j].SeatID
	})

	return seats[0].SeatID
}
